sound/soc/intel/boards/skl_hda_dsp_generic.c

```diff
@@ -135,11 +138,20 @@ static int skl_hda_fill_card_info(struct snd_soc_acpi_mach_params *mach_params)
 				skl_hda_be_dai_links[IDISP_DAI_COUNT +
 					HDAC_DAI_COUNT + i];
 		}
-	} else if (codec_count == 2 && codec_mask & IDISP_CODEC_MASK) {
+	} else if (codec_count <= 2) {
 		num_links = ARRAY_SIZE(skl_hda_be_dai_links);
 		num_route = ARRAY_SIZE(skl_hda_map);
 		card->dapm_widgets = skl_hda_widgets;
 		card->num_dapm_widgets = ARRAY_SIZE(skl_hda_widgets);
+		if (!(codec_mask & IDISP_CODEC_MASK)) {
+			if (codec_count == 2)
+				return -EINVAL;
+			for (i = 0; i < IDISP_DAI_COUNT; i++) {
+				skl_hda_be_dai_links[i].codecs = dummy_codec;
```

Comment:
I'm a bit confused now, didn't you have
```
skl_hda_be_dai_links[i].codecs->name = ...;
```
before? That would mean, that there was at least one element assigned to `.codecs`. So, now you're overwriting that? I.e. you're relying on the fact, that the original `.codecs` was allocated with `devm_kcalloc()`, so, it will be freed anyway?

Reply:
@lyakh Yeah, trying to imitatte the best modern ASoC DAI macro usages here. :) So the codecs fields are typically filled with the SND_SOC_DAILINK_() macros and that also applies to this machine driver. These macros create and fill static structs in the compilation unit. So be default the codecs field is a pointer to a static struct declated in skl_hda_dsp_common.c . What I add in this patch is another SND_SOC_DALINK_DEF that creates static struct 'dummy_codec' and in this loop I changes codecs to point to it. Based on my understanding and review of the macros, this is safe as codecs always points to one of the statically allocated structs and never to a dynamically allocated object. This code is local to the generic hda driver, so other entities will not modify skl_hda_be_dai_links directly.

But yeah, scrutiny of the logic is much welcome.